### ISSUE TITLE
CredSSP does not work

### ISSUE THREAD

**Issue by evg-zhabotinsky** [rdesktop/rdesktop#71]
```
rdesktop host
ERROR: CredSSP: Initialize failed, do you have correct kerberos tgt initialized ?
Failed to connect, CredSSP required by server.
```

When I login using MS client (from both Android and Windows), everything just works.
Maybe I just don't understand how to configure this properly, but then I suppose it has to be explained better in the man page, because moderate amount of googling didn't really help me.

In this specific case, I can disable network level authentication (what google suggests), but it's not really a solution.

**Comment by hean01-cendio** [comment]
rdesktop supports CredSSP with kerberos and lack the implementation of NTLM. That's why there are a differences from MS client and rdesktop. To get this working, you need to configure kerberos on the client side against MS KDC (Active directory service) to be able to obtain a ticket used for NLA.

I added issue #72 for the lack of NTLM.

**Comment by hean01-cendio** [comment]
On this issue, we should look into adding information about how this works, maybe a section on our web site.

**Comment by flip111** [comment]
Is this about automatic login on windows? I noticed when i login to windows from windows that i don't have to fill in my credentials. With rdesktop i have to enter my user name and password as if i was sitting at the pc itself.

**Comment by hean01-cendio** [comment]
> Is this about automatic login on windows? 

Yes it is. On a windows client you have a credential cache that is used to authenticate against services in your network (NTLM). RDP have support for such "Single-Sign-On" authentication, however Linux clients does not have this cached authentication and there for  you need to provide credentials, either via rdesktop command line or in login window at remote session.

However, there is something called Kerberos which is an authentication mechanism for requesting access to services based on a initial "ticket".  Windows AD (Active Directory) provides a Kerberos infrastructure and a Linux box can be configure to authenticate against that. This means that when you login to you linux box, you will authenticate for a initial  kerberos "ticket" that is used to access other services such as RDP. Then when running rdesktop, CredSSP will check if you have a ticket for accessing the remote service and use that for authentication "Singel-Sign-On" against the remote RDS server. If there is not ticket, rdesktop will fallback to plain TLS connection.

**Comment by HenrikBach1** [comment]
Hi,

As long this issue isn't fixed (by documentation and/or code) no one (or at least I :-) ) is not able to login into Microsoft Azure provisioned Windows Server 2016 (v1607) virtual machines with rdesktop.

I notice, that freerdp (on centos 7, somehow) is able to connect with above virtual machine. However, freerdp isn't behaving nicely not keeping meta-keypresses inside the client, even though it is setup to do so, by default.

/Henrik

**Comment by gp54321** [comment]
Henrik, I can connect to a Windows server 2016 build 1607 using rdesktop 1.8.3 (this is not on Azure though, maybe it could have an effect on the Kerberos port ? the port 88 was used for this test, I have no idea how it can work with Azure). The idea is pretty simple: install krb5-user, setup /etc/krb5.conf, launch rdesktop

If my windows username is billg, the win domain is contoso.local, you have to set up /etc/krb5.conf with something like that:

[libdefaults]
 default_realm = CONTOSO.LOCAL
[realms]
 CONTOSO.LOCAL = {
  kdc = 10.1.0.5:88
  admin_server = 10.1.0.5:749
  default_domain = CONTOSO.LOCAL
 }
[domain_realm]
 .contoso.local = CONTOSO.LOCAL

then (no need to reboot or anything)

kinit billg
<enter billg password>
rdesktop -u billg -d CONTOSO.LOCAL -f MYSERVER
with MYSERVER defined in /etc/hosts
This has been tested with a Mandriva 2010 distro from the Mesozoic age and a more modern Ubuntu 16.04 LTS, it worked basically the same (a bit easier with Ubuntu)
The only trouble is that I suffer from a double login: kinit is basically a login, but the server connection screen complains that the password is invalid; if I enter again the password at this stage, I am good to go.
Please note that kinit should not return any error. If you get an error about an unexpected reply and it works with -C option, you have probably an upper/lower case problem with the domain name. 
Also do not use an ip address on the rdesktop call, use a symbolic name like I have shown you otherwise it will not work.

Last thing: rdesktop is not very user-friendly for windows errors: first try I did, the user was not 
allowed to connect, and rdesktop just terminated silently after saying 'connection established using CredSSP' , it was only after a connection was attempted on a Windows computer that I was allowed to bang my head against the keyboard.

Good luck ! it's not really rocket science.

**Comment by hean01-cendio** [comment]
@gp54321 Thanks for your response, please if you do stumble upon any fishy with Window 2016 server, report it and link to issue #101, TIA.

**Comment by ppokhriyal** [comment]
CREDSSP WORKING [ERROR: CredSSP: Initialize failed, do you have correct kerberos tgt initialized ? ]

1 ) First create the /etc/krb5.conf 
  [libdefaults]
  default_realm = EXAMPLE.COM
 dns_lookup_realm = false
 dns_lookup_kdc = false
 ticket_lifetime = 24h
 forwardable = yes
[realms]
 EXAMPLE.COM = {
admin_server = example.com
default_domain = example.com
kdc = example.com
}
vdi.com = EXAMPLE.COM
.vdi.com = EXAMPLE.COM

2) Make sure you have /etc/gssapi_mech.conf [which searches for libgssapi_krb5.so.2 ]
3) kinit USERNAME_OF_ABC.EXAMPLE.COM
4) Above command will create a temporary file in /tmp/krb5cc_0
5) Now run rdesktop ABC.EXAMPLE.COM

**Comment by jlinkels** [comment]
@ppokhriyal: thanks for posting. But since "domain names" are ambiguous in a Windows environment I can't grasp the understanding of what you write.

I have a local Windows 2k12 server on the network and the Linux client should connect to it. 

The login domain is 2k12-DC. The server is local on the LAN, no connection to the outside.

The FQDN of the 2k12 server is 2k12-DC.2k12test.local. So I assume the domain is 2k12test.local. (At least according to the 2k12 DNS server)

The Linux client running rdesktop does not use the Windows 2k12 box as DNS server. It is not connected to the Windows domain or AD at all.

On the linux client I can edit the hosts file. The DNS is normally pointing to the internet router. Not to the 2k12 box. Should I add entries in the hosts file so it can find 2k12test.local? 

Summarizing, if I log in as 2k12-DC\Administrator to 2k12-DC.2k12test.local how should I compose the krb5.conf file?

jlinkels

**Comment by ppokhriyal** [comment]
@jlinkels : In my environment i am trying to take rdesktop from my customize Linux machine to a Windows machine which is a member of a domain example:vdi.com

So, when i fire rdesktop command from my linux machine
 example : rdesktop WIN12.vdi.com
i get this output  >>ERROR: CredSSP: Initialize failed, do you have correct kerberos tgt initialized ? 

So, in order to resolve this Error i have followed the below steps 

###My apologizes i didn't mentioned the ADS part ####

Domain: vdi.com
Hostname:xyz
  
1 . First you have to join your Linux Machine to the DOMAIN in my case  it is vdi.com

    Steps for NET ADS JOIN

    a) First  edit the krb.conf file

        [libdefaults]
        default_realm = VDI.COM
        dns_lookup_realm = false
        dns_lookup_kdc = false
        ticket_lifetime = 24h
        forwardable = yes
      [realms]
        VDI.COM = {
        admin_server = vdi.com
        default_domain = vdi.com
        kdc = vdi.com
      }
    vdi.com = VDI.COM
   .vdi.com = VDI.COM

b) /etc/host

    IP-ADDRESS   xyz.vdi.com xyz localhost
    IP-ADDRESS   xyz xyz.vdi.com

c) /etc/samba/smb.conf

    #START# 
    # Global parameters
    [global]
    unix charset = ISO-8859-1
    workgroup = VDI
    realm = VDI.COM
    winbind separator = +
    security = user
    winbind use default domain = yes
    server string = WORKGROUP
    winbind enum users = Yes
    winbind enum groups = Yes
    winbind refresh tickets = Yes
    winbind offline logon = Yes
    winbind normalize names = Yes
    idmap config DOMAIN:range = 10000-999999
    idmap config DOMAIN:schema_mode = rfc2307
    idmap config DOMAIN:backend = ad
    idmap config *:range = 2000-9999
    idmap config * : backend = tdb
    ldap ssl = No
    read only = No
    create mask = 0777
    directory mask = 0777
    client signing = mandatory
    server signing = mandatory
    guest account = nobody
   map to guest = Never
   restrict anonymous = 2
   [smb-share] 
   path = /tmp/smb-share
   read only = No
   guest ok = Yes
   public = Yes
   browseable = Yes 
   writable = Yes
   create mode = 0700
#END#

d) restart the SAMBA and NMBD service    
e) net ads join -Uusername%password
f)  After successful domain join run kinit USERNAME_OF_VDI.COM then it will ask for password.

2. rdesktop 2k12-DC.2k12test.local

I hope this solution works for you.

**Comment by jlinkels** [comment]
@ppokhriyal: Thanks for taking the time to explain it. It works using the GUI remote desktop client.
```
Windows server:
Computer name: 2k12-dc
Domain: 2k12test.local

This is what my krb5.conf looks:
```
[libdefaults]
	default_realm = 2K12TEST.LOCAL
	dns_lookup_realm = false
	dns_lookup_kdc = false
	ticket_lifetime = 24h
	forwardable = yes


[realms]
	2K12TEST.LOCAL = {
		kdc = **2k12-dc**.2k12test.local
		default_domain = 2k12test.local
		admin_server = **2k12-dc**.2k12test.local
	}
2k12test.local = 2K12TEST.LOCAL
.2k12test.local = 2K12TEST.LOCAL
```
Then the /etc/hosts file: you proposed this:
```
IP-ADDRESS   xyz.vdi.com xyz localhost
IP-ADDRESS   xyz xyz.vdi.com
```
But it is not clear if you intend to create a record for the host to connect or a record for localhost. Anyway, my /etc/hosts looks like this:
```
127.0.0.1	localhost
127.0.1.1	lampje-pc.2k12test.local	lampje-pc
192.168.110.11	2k12-DC 2k12-DC.2k12test.local
```
lampje-pc is the client. But it is not on the domain 2k12-DC.2k12test.local. 

Very important: after configuring krb5.conf, run:
kinit <username on windows computer>

Anyway, in the GUI remote desktop client I have to enter:
Computer: 2k12-dc.2k12test.local (or 192.168.110.11)
Username: Administrator
Password: <admin password>
Domain: 2k12-dc (yeah right: the computer name!)
And then select Windows XP/2003

For that matter this works as well:
rdesktop -u administrator -p <admin password> -f 2k12-dc.2k12test.local
Hit CTRL-ALT-ENTER to toggle full screen.

Thanks again

**Comment by hifi** [comment]
Hello, I get the CredSSP error when connecting to a remote Windows Server 2012R2.

However, I only have a TCP port forwarding for the RDP port of the server and I can successfully connect to it from Windows 10. I don't need to configure anything on Windows 10 to connect to the server, only provide the domain along with the username and it just works. It's not connected to any domain, it's just a regular standalone installation.

Is this an unrelated issue because setting these kerberos things and ips in /etc/hosts wouldn't really do anything good when I can only connect to a tunneled TCP 3389, right?

It seems to me the client would only need to create this ticket locally for the connection to succeed or somehow gracefully fail SSP so a normal login would work?

**Comment by hean01-cendio** [comment]
@hifi, rdesktop only support CredSSP with kerberos, mstsc uses CredSSP with NTLM and that is why it works out of the box. See bug #72 for adding NTLM support to rdesktop. And you are right about the kerberos part, that needs a infrastructure and be accessible from the client side of firewall, eg. no single port setup there.

**Comment by hifi** [comment]
@hean01 Thank you for clearing that up, I'll subscribe to that issue.

**Comment by DesktopECHO** [comment]
Earlier this year I was annoyed to discover I couldn't connect to a newly-provisioned Azure VM from my iBook G4, so I compiled rdesktop 1.7 with a [CredSSP+NTLMv2 patch from way back in 2011](http://opensource.dell.com/releases/Dell_Software/rdesktop-rdpv6-credssp-readme.html) and was able to login to my new VM without needing to edit krb5.conf or smb.conf.  

It would be nice to see this patch in newer versions of rdesktop but in my case the libraries on Mac OS 10.4 were too old to make any headway.

**Comment by sebma** [comment]
Hi everyone, why is `xfreerdp` working for this kind of connection ?

**Comment by MarkLTZ** [comment]
Hello! Any progress using CredSSp over NTLM without the need of Kerberos?
Thanks

**Comment by thomas725** [comment]
Is there a dummy guide how to get this working with rdesktop? I have no idea what to do with the hints about `/etc/krb5.conf` and `/etc/gssapi_mech.conf` given above. Can somebody dumb it down for me?

This works fine out of the box with Remmina, but I'd prefer the mostly gui-less rdesktop, which fails me for some target machines like this:
```
$ rdesktop -k de -r clipboard:CLIPBOARD -u user -p password -g 1916x1025 -v target-ip
is_wm_active(): WM name: KWin
Connecting to server using NLA...
Core(warning): Certificate received from server is NOT trusted by this system, an exception has been added by the user to trust this specific certificate.
TLS  Session info: (TLS1.2)-(RSA)-(AES-256-GCM)

Failed to initialize NLA, do you have correct Kerberos TGT initialized ?
Failed to connect using NLA, trying with SSL
Failed to connect, CredSSP required by server (check if server has disabled old TLS versions, if yes use -V option).
```

**Comment by bemonolit** [comment]
xfreerdp /u:"abel.com\adam" /v:192.168.1.113:3389   works fine!!!
